docs/log-levels.md

```diff
@@ -22,17 +22,17 @@ In the following section we'll explore how to use them in practice.
 
 It is always fine for a library to log at `trace` and `debug` levels, and these two should be the primary levels any library is logging at.
 
-`trace` is the finest log level, and end-users of a library will not usually use it unless debugging very specific issues. You should consider it as a way for library developers to "log everything we could possibly need to diagnose a hard to reproduce bug." It is expected to take a toll on the performance of a system, and developers can assume trace level logging will not be used in production deployments, unless enabled specifically to locate some specific issue.
+`trace` is the finest log level, and end-users of a library will not usually use it unless debugging very specific issues. You should consider it as a way for library developers to "log everything we could possibly need to diagnose a hard to reproduce bug."  Unrestricted logging at `trace` level may take a toll on the performance of a system, and developers can assume trace level logging will not be used in production deployments, unless enabled specifically to locate some specific issue.
 
 This is in contrast with `debug` which some users _may_ choose to run enabled on their production systems.
 
 > Debug level logging should be not "too" noisy. Developers should assume some production deployments may need to (or want to) run with debug level logging enabled. 
 >
 > Debug level logging should not completely undermine the performance of a production system.
```

Comment:
This very much depends on a number of factors: the appender used, the volume of data, the general retention of the data and the overall propagation of the logs.

For example, a debug system logging to an in-memory ring buffer may have very low overhead even at high volume, while debug logging for all servers sent to centralized logging which is indexed and retained by ELK or Splunk may have a high cost and may overwhelm the indexing resources of the logging system.

It may be worth defining the limits of debug logging, i.e. should not saturate local CPU/IO/Network resources of the system and state that centralized logging may have additional variables to take into account.

Reply:
Yeah that's fair but somewhat of a logging backend topic... I think it is right to say in any case "logging on debug level should not overwhelm systems" because there isn't really a good smart ( like you mention with ring buffering debug messages etc) backend around for logging in swift, and people use random backends (just println based ones...)

>  Marick deals with some of this, but there are probably dedicated Apple reporting tools for capturing and retrieving debug/trace logs when you don't have access.

Yeah that's cool, though that would be about specific appenders (backends how we call them).
The guide here only speaks in terms of the logging API -- think "slf4j-api" so it can't really assume much.

-- 

os_log is outside of the realm that this doc speaks about, I'll clarify this perhaps.

This is the "swift server guides" and as such there is no os_log because that's a darwin specific API.